homeassistant/components/zha/core/listeners.py

```diff
@@ -27,90 +55,439 @@ def parse_and_log_command(entity_id, cluster, tsn, command_id, args):
 
 
 class LevelListener(ClusterListener):
     """Listener for the LevelControl Zigbee cluster."""
 
+    name = LEVEL
+
     CURRENT_LEVEL = 0
 
+    @callback
     def cluster_command(self, tsn, command_id, args):
         """Handle commands received to this cluster."""
         cmd = parse_and_log_command(
-            self._entity.entity_id,
+            self.unique_id,
             self._cluster,
             tsn,
             command_id,
             args
         )
 
         if cmd in ('move_to_level', 'move_to_level_with_on_off'):
-            self._entity.set_level(args[0])
+            self.dispatch_level_change(SIGNAL_SET_LEVEL, args[0])
         elif cmd in ('move', 'move_with_on_off'):
             # We should dim slowly -- for now, just step once
             rate = args[1]
             if args[0] == 0xff:
                 rate = 10  # Should read default move rate
-            self._entity.move_level(-rate if args[0] else rate)
+            self.dispatch_level_change(
+                SIGNAL_MOVE_LEVEL, -rate if args[0] else rate)
         elif cmd in ('step', 'step_with_on_off'):
             # Step (technically may change on/off)
-            self._entity.move_level(-args[1] if args[0] else args[1])
+            self.dispatch_level_change(
+                SIGNAL_MOVE_LEVEL, -args[1] if args[0] else args[1])
 
+    @callback
     def attribute_updated(self, attrid, value):
         """Handle attribute updates on this cluster."""
+        _LOGGER.debug("%s: received attribute: %s update with value: %i",
+                      self.unique_id, attrid, value)
         if attrid == self.CURRENT_LEVEL:
-            self._entity.set_level(value)
+            self.dispatch_level_change(SIGNAL_SET_LEVEL, value)
+
+    def dispatch_level_change(self, command, level):
+        """Dispatch level change."""
+        async_dispatcher_send(
+            self._zha_device.hass,
+            "{}_{}".format(self.unique_id, command),
+            level
+        )
+
+    async def async_initialize(self, from_cache):
+        """Initialize listener."""
+        await self.get_attribute_value(
+            self.CURRENT_LEVEL, from_cache=from_cache)
+        await super().async_initialize(from_cache)
+
+
+class IASZoneListener(ClusterListener):
+    """Listener for the IASZone Zigbee cluster."""
+
+    name = 'zone'
+
+    @callback
+    def cluster_command(self, tsn, command_id, args):
+        """Handle commands received to this cluster."""
+        if command_id == 0:
+            state = args[0] & 3
+            async_dispatcher_send(
+                self._zha_device.hass,
+                "{}_{}".format(self.unique_id, SIGNAL_ATTR_UPDATED),
+                state
+            )
+            _LOGGER.debug("Updated alarm state: %s", state)
+        elif command_id == 1:
+            _LOGGER.debug("Enroll requested")
+            res = self._cluster.enroll_response(0, 0)
+            self._zha_device.hass.async_add_job(res)
```

Comment:
`hass.async_add_job` is legacy. Use `hass.async_create_task` or `hass.async_add_executor_job` as appropriate.

Reply:
Use `hass.async_add_executor_job` when we need to schedule regular I/O on the executor thread pool. Use `hass.async_create_task` when we need to schedule an asyncio task on the event loop.

Reply:
Used create task